benchmarks/movielens.ipynb

```diff
@@ -0,0 +1,942 @@
+      "    \"Recall@k\": 0.015325076842881835,\n",
+      "    \"nDCG@k\": 0.034274974910292674\n",
+      "}\n",
+      "\n",
+      "Computing ncf algorithm on Movielens 1m\n",
+      "Training model: neumf\n",
+      "Training time: 0:04:03.736511\n",
+      "\n",
+      "Evaluating with ncf\n",
+      "Ranking prediction time: 0:00:38.344802\n",
+      "Ranking metrics: \n",
+      "{\n",
+      "    \"MAP\": 0.053581743438381235,\n",
+      "    \"Precision@k\": 0.2892827563359285,\n",
+      "    \"Recall@k\": 0.09361653852778785,\n",
+      "    \"nDCG@k\": 0.3161943185002138\n",
+      "}\n",
+      "\n",
+      "Computing fastai algorithm on Movielens 1m\n",
+      "Training time: 0:03:21.684542\n",
+      "\n",
+      "Evaluating with fastai\n",
+      "Rating prediction time: 0:00:00.421794\n",
+      "Rating metrics: \n",
+      "{\n",
+      "    \"Explained Variance\": 0.3789675628236845,\n",
+      "    \"MAE\": 0.7017161939976656,\n",
+      "    \"R2\": 0.3763750984206222,\n",
+      "    \"RMSE\": 0.8819809932848574\n",
+      "}\n",
+      "Ranking prediction time: 0:00:52.389096\n",
+      "Ranking metrics: \n",
+      "{\n",
+      "    \"MAP\": 0.02437502428643567,\n",
+      "    \"Precision@k\": 0.16029484843465297,\n",
+      "    \"Recall@k\": 0.053431011482860266,\n",
+      "    \"nDCG@k\": 0.1751913542471727\n",
+      "}\n",
+      "CPU times: user 20min 10s, sys: 1min 11s, total: 21min 22s\n",
+      "Wall time: 24min 7s\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%time\n",
+    "\n",
+    "# For each data size and each algorithm, a recommender is evaluated. \n",
+    "cols = [\"Data\", \"Algo\", \"K\", \"Train time\", \"Rating time\", \"RMSE\", \"MAE\", \"R2\", \"Explained Variance\", \"Ranking time\", \"MAP\", \"nDCG@k\", \"Precision@k\", \"Recall@k\", ]\n",
+    "df_results = pd.DataFrame(columns=cols)\n",
+    "\n",
+    "for data_size in data_sizes:\n",
+    "    # Load the dataset\n",
+    "    df = movielens.load_pandas_df(\n",
+    "        size=data_size,\n",
+    "        header=[USER_COL, ITEM_COL, RATING_COL, TIMESTAMP_COL]\n",
+    "    )\n",
+    "    print(\"Size of Movielens {}: {}\".format(data_size, df.shape))\n",
+    "    \n",
+    "    # Split the dataset\n",
+    "    df_train, df_test = python_stratified_split(df,\n",
+    "                                                ratio=0.75, \n",
+    "                                                min_rating=1, \n",
+    "                                                filter_by=\"item\", \n",
+    "                                                col_user=USER_COL, \n",
+    "                                                col_item=ITEM_COL\n",
+    "                                                )\n",
+    "    print(\"Train set size: {}\".format(df_train.shape))\n",
+    "    print(\"Test set size: {}\".format(df_test.shape))\n",
+    "   \n",
+    "    # Loop through the algos\n",
+    "    for algo in algorithms:\n",
```

Comment:
Each iteration takes long and all are independent, which is a perfect scenario for parallel / distributed processing. Can we use Azure remote compute for this or (if refactoring is not trivial or already in the next todo list), maybe simply use multiprocessing at least for CPU algos?

Reply:
yeah I think the next step on this would be to do that, if we do the benchmarks through notebooks we can do them in parallel. It would be nice. We also should start thinking on how to perform the tests in parallel, which is blocking us a little bit